packages/js/src/plugins/auctionHouseModule/operations/directBuy.ts

```diff
@@ -0,0 +1,251 @@
+// Operation
+// -----------------
+
+const Key = 'DirectBuyOperation' as const;
+
+/**
+ * Creates a bid on a given asset and then executes a sale on the created bid and listing.
+ *
+ * ```ts
+ * await metaplex
+ *   .auctionHouse()
+ *   .buy({ auctionHouse, listing, buyer })
+ *   .run();
+ * ```
+ *
+ * @group Operations
+ * @category Constructors
+ */
+export const directBuyOperation = useOperation<DirectBuyOperation>(Key);
+
+/**
+ * @group Operations
+ * @category Types
+ */
+export type DirectBuyOperation = Operation<
+  typeof Key,
+  DirectBuyInput,
+  DirectBuyOutput
+>;
+
+/**
+ * @group Operations
+ * @category Inputs
+ */
+export type DirectBuyInput = {
+  /** The Auction House in which to create a Bid and execute a Sale. */
+  auctionHouse: AuctionHouse;
+
+  /**
+   * The Auction House authority.
+   * If this is Signer the transaction fee
+   * will be paid from the Auction House Fee Account
+   *
+   * @defaultValue `auctionHouse.authority`
+   */
+  authority?: PublicKey | Signer;
+
+  /**
+   * Creator of a bid.
+   *
+   * @defaultValue `metaplex.identity()`
+   */
+  buyer?: Signer;
+
+  /**
+   * The Listing that is used in the sale.
+   * We only need a subset of the `Listing` model but we
+   * need enough information regarding its settings to know how
+   * to execute the sale.
+   *
+   * This includes, its asset, auction house address, seller, receipt address etc.
+   */
+  listing: Pick<
+    Listing,
+    | 'asset'
+    | 'auctionHouse'
+    | 'canceledAt'
+    | 'price'
+    | 'sellerAddress'
+    | 'tokens'
+    | 'tradeStateAddress'
+    | 'receiptAddress'
+  >;
+
+  /**
+   * The Auctioneer authority key.
+   * It is required when Auction House has Auctioneer enabled.
+   *
+   * @defaultValue No default value.
+   */
+  auctioneerAuthority?: Signer;
+
+  /**
+   * The address of the bookkeeper wallet responsible for the receipt.
+   *
+   * @defaultValue `metaplex.identity()`
+   */
+  bookkeeper?: Signer;
+
+  /** A set of options to configure how the transaction is sent and confirmed. */
+  confirmOptions?: ConfirmOptions;
+};
+
+/**
+ * @group Operations
+ * @category Outputs
+ */
+export type DirectBuyOutput = {
+  /** A model that keeps information about the Bid. */
+  bid: Bid;
+
+  /** A model that keeps information about the Purchase. */
+  purchase: Purchase;
+
+  /** The blockchain response from sending and confirming the transaction. */
+  response: SendAndConfirmTransactionResponse;
+};
+
+/**
+ * @group Operations
+ * @category Handlers
+ */
+export const directBuyOperationHandler: OperationHandler<DirectBuyOperation> = {
+  handle: async (operation: DirectBuyOperation, metaplex: Metaplex) => {
+    const { auctionHouse } = operation.input;
+    const { receipt, bid, response } = await directBuyBuilder(
+      metaplex,
+      operation.input
+    ).then((buyBuilder) =>
+      buyBuilder.sendAndConfirm(metaplex, operation.input.confirmOptions)
+    );
+
+    const purchase = await metaplex
```

Comment:
It won't work if the listing doesn't have a receipt, you can check executeSale.

const shouldPrintReceipt =
    (params.printReceipt ?? true) &&
    Boolean(**listing.receiptAddress && bid.receiptAddress** && !isPartialSale);